### PR TITLE
Add initial IPA-Tuura federation

### LINKED ISSUE #35325: Experimental support for ipa-tuura user federation
### Description

Add experimental support for the IPA-Tuura federation in Keycloak.

IPA-Tuura is a bridge service to access the identity providers (FreeIPA/AD/LDAP). It is based on the SCIMv2 API and provides an improved integration for classic directory services to cloud environments with an easy deployment. The bridge can also be used in a variety of different scenarios, from migration to synchronization of identities across different providers.

### PR DESCRIPTION
**Pull request by sguilhen**
Closes #35325

### CODE REVIEW

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        return response;
+    }
+
+    private void setUserAttr(SCIMUser.Resource user, String attr, String value) {
+        SCIMUser.Resource.Name name = user.getName();
+        SCIMUser.Resource.Email email = new SCIMUser.Resource.Email();
+        List<SCIMUser.Resource.Email> emails = new ArrayList<SCIMUser.Resource.Email>();
+
+        switch (attr) {
+            case "firstName":
```
Comment (pedroigor) [inline federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java:329]: It is a nitpick but we are fetching root attribute names names from constants in `UserModel`. Mainly in user profile, where we usually want to see where those attributes are being used.

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+         */
+        int idx = username.indexOf("@");
+        if (idx != -1) {
+            username = username.substring(0, idx);
+        }
+
+        UserModel user = UserStoragePrivateUtil.userLocalStorage(session).getUserByUsername(realm, username);
+        if (user != null) {
+            logger.debug("User already exists in keycloak");
+            return user;
```
Comment (pedroigor) [inline federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java:107]: We should probably consider in the next iteration to return a delegate that tracks state changes to sync them back to the SCIM service provider in a transaction.

Btw, do we want to return `IpatuuraUserModelDelegate` instead when the local account exists?

File: federation/ipatuura/pom.xml [hunk 3]
```diff
@@ -0,0 +1,64 @@
+<!--
+  ~ Copyright 2016 Red Hat, Inc. and/or its affiliates
```
Comment (pedroigor) [inline federation/ipatuura/pom.xml:2]:
```suggestion
  ~ Copyright 2024 Red Hat, Inc. and/or its affiliates
```

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProviderFactory.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    public void validateConfiguration(KeycloakSession session, RealmModel realm, ComponentModel config)
+            throws ComponentValidationException {
+        Ipatuura ipatuura = new Ipatuura(session, config);
+
+        SimpleHttp.Response response;
+
+        try {
+            response = ipatuura.clientRequest("", "GET", null);
```
Comment (pedroigor) [inline federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProviderFactory.java:78]: As a follow-up, we can consider adding something similar to LDAP where you test both the connection and authentication, separately.

**Review by pedroigor — COMMENTED** [review]
LGTM. Only a few comments. Mainly follow-ups.

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java (hunk 1 above)
Comment (sguilhen) [reply federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java:329]: You mean we should be using the constants here? I can fix that.

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProviderFactory.java (hunk 4 above)
Comment (sguilhen) [reply federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProviderFactory.java:78]: Yeah, but that will require addition of specific UI for this provider along with a rest endpoint for testing the connection. Definitely worth doing it, just a completely different feature.

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java (hunk 2 above)
Comment (sguilhen) [reply federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java:107]: I'm not sure if the code ever reaches this - the call `getUserByUsername` goes through the `UserStorageManager`, which checks the local storage and calls `validate` on the provider. This method is returning the delegate already.

Comment (sguilhen) [reply federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java:107]: So in a way the provider is repeating the local storage check needlessly

File: federation/ipatuura/pom.xml (hunk 3 above)
Comment (sguilhen) [reply federation/ipatuura/pom.xml]: Done!

File: federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java (hunk 1 above)
Comment (sguilhen) [reply federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/Ipatuura.java]: Done!

**Review by pedroigor — APPROVED** [review]

**Pull request merged**